stan/math/rev/fun/tcrossprod.hpp

```diff
@@ -21,41 +21,26 @@ namespace math {
  * @param M Matrix to multiply.
  * @return M times its transpose.
  */
-template <typename T, require_eigen_vt<is_var, T>* = nullptr>
-inline Eigen::Matrix<var, T::RowsAtCompileTime, T::RowsAtCompileTime>
-tcrossprod(const T& M) {
-  if (M.rows() == 0) {
-    return {};
-  }
-  // if (M.rows() == 1)
-  //   return M * M.transpose();
-
-  // WAS JUST THIS
-  // matrix_v result(M.rows(), M.rows());
-  // return result.setZero().selfadjointView<Eigen::Upper>().rankUpdate(M);
-
-  Eigen::Matrix<var, T::RowsAtCompileTime, T::RowsAtCompileTime> MMt(M.rows(),
-                                                                     M.rows());
-
-  vari** vs
-      = reinterpret_cast<vari**>(ChainableStack::instance_->memalloc_.alloc(
-          (M.rows() * M.cols()) * sizeof(vari*)));
-  int pos = 0;
-  for (int m = 0; m < M.rows(); ++m) {
-    for (int n = 0; n < M.cols(); ++n) {
-      vs[pos++] = M(m, n).vi_;
-    }
-  }
-  for (int m = 0; m < M.rows(); ++m) {
-    MMt.coeffRef(m, m)
-        = var(new internal::dot_self_vari(vs + m * M.cols(), M.cols()));
-  }
-  for (int m = 0; m < M.rows(); ++m) {
-    for (int n = 0; n < m; ++n) {
-      MMt.coeffRef(n, m) = MMt.coeffRef(m, n) = dot_product(M.row(m), M.row(n));
-    }
+template <typename T, require_rev_matrix_t<T>* = nullptr>
+inline plain_type_t<T> tcrossprod(const T& M) {
+  using T_double = Eigen::MatrixXd;
+  using T_var = plain_type_t<T>;
+
+  if (M.size() == 0) {
+    return M;
   }
-  return MMt;
+
+  arena_t<plain_type_t<T>> arena_M = M;
+  arena_t<Eigen::MatrixXd> arena_M_val = value_of(arena_M);
+
+  arena_t<T_var> res = arena_M_val * arena_M_val.transpose();
+
+  reverse_pass_callback([res, arena_M, arena_M_val]() mutable {
+    Eigen::MatrixXd adj = res.adj();
```

Comment:
```suggestion
    const auto& adj = to_ref(res.adj());
```
This way we can avoid the copy in the varmat case.